hudi-io/src/main/java/org/apache/hudi/io/hfile/HFileReader.java

```diff
@@ -0,0 +1,178 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.hudi.io.hfile;
+
+import org.apache.hadoop.fs.FSDataInputStream;
+
+import java.io.ByteArrayInputStream;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * A reader reading a HFile.
+ */
+public class HFileReader {
+  private final FSDataInputStream stream;
+  private final long fileSize;
+  private boolean isMetadataInitialized = false;
+  private HFileContext context;
+  private List<BlockIndexEntry> blockIndexEntryList;
+  private HFileBlock metaIndexBlock;
+  private HFileBlock fileInfoBlock;
+
+  public HFileReader(FSDataInputStream stream, long fileSize) {
+    this.stream = stream;
+    this.fileSize = fileSize;
+  }
+
+  /**
+   * Initializes the metadata by reading the "Load-on-open" section.
+   *
+   * @throws IOException upon error.
+   */
+  public void initializeMetadata() throws IOException {
+    assert !this.isMetadataInitialized;
+
+    // Read Trailer (serialized in Proto)
+    HFileTrailer trailer = readTrailer(stream, fileSize);
+    this.context = HFileContext.builder()
+        .compressAlgo(trailer.getCompressionCodec())
+        .build();
+    HFileBlockReader blockReader = new HFileBlockReader(
+        context, stream, trailer.getLoadOnOpenDataOffset(), fileSize - HFileTrailer.getTrailerSize());
+    HFileRootIndexBlock dataIndexBlock =
+        (HFileRootIndexBlock) blockReader.nextBlock(HFileBlockType.ROOT_INDEX);
+    this.blockIndexEntryList = dataIndexBlock.readDataIndex(trailer.getDataIndexCount());
+    this.metaIndexBlock = blockReader.nextBlock(HFileBlockType.ROOT_INDEX);
+    this.fileInfoBlock = blockReader.nextBlock(HFileBlockType.FILE_INFO);
+
+    this.isMetadataInitialized = true;
+  }
+
+  /**
+   * Seeks to the key to look up.
+   *
+   * @param key Key to look up.
+   * @return The {@link KeyValue} instance in the block that contains the exact same key as the
+   * lookup key; or empty {@link Optional} if the lookup key does not exist.
+   * @throws IOException upon error.
+   */
+  public Optional<KeyValue> seekTo(Key key) throws IOException {
+    BlockIndexEntry lookUpKey = new BlockIndexEntry(key, -1, -1);
+    int rootLevelBlockIndex = searchBlockByKey(lookUpKey);
+    if (rootLevelBlockIndex < 0) {
+      // Key smaller than the start key of the first block
+      return Optional.empty();
+    }
+    BlockIndexEntry blockToRead = blockIndexEntryList.get(rootLevelBlockIndex);
+    HFileBlockReader blockReader = new HFileBlockReader(
+        context, stream, blockToRead.getOffset(), blockToRead.getOffset() + (long) blockToRead.getSize());
+    HFileDataBlock dataBlock = (HFileDataBlock) blockReader.nextBlock(HFileBlockType.DATA);
```

Comment:
we should return multiple values per search? for e.g col stats, we can do a prefix search and it can go over several data blocks?

Reply:
More APIs need to be added to the HFile reader based on the current prefix search implemented in `HoodieAvroHFileReader` (https://github.com/apache/hudi/blob/release-0.14.0/hudi-common/src/main/java/org/apache/hudi/io/storage/HoodieAvroHFileReader.java#L298). I plan to cover this in a separate PR: HUDI-7217

Reply:
Disregard the last comment.  I've supported multiple `seekTo` calls from the same reader, supporting prefix search too.